services/src/main/java/edu/unc/lib/dl/cdr/services/rest/DatastreamRestController.java

```diff
@@ -15,57 +15,88 @@
-import edu.unc.lib.dl.ui.util.AnalyticsTrackerUtil.AnalyticsUserData;
-import edu.unc.lib.dl.util.ContentModelHelper;
+import edu.unc.lib.dl.ui.util.AnalyticsTrackerUtil;
 
 /**
- * 
+ *
  * @author bbpennel
  *
  */
 @Controller
 public class DatastreamRestController {
+    private static final Logger log = LoggerFactory.getLogger(DatastreamRestController.class);
+
     @Autowired
     private FedoraContentService fedoraContentService;
+    @Autowired
+    private AnalyticsTrackerUtil analyticsTracker;
 
     @RequestMapping("/file/{pid}")
-    public void getDatastream(@PathVariable("pid") String pid,
+    public void getDatastream(@PathVariable("pid") String pidString,
             @RequestParam(value = "dl", defaultValue = "false") boolean download,
-            HttpServletRequest request, HttpServletResponse response) {
-        fedoraContentService.streamData(pid, null, download, new AnalyticsUserData(request), response);
+            HttpServletRequest request,
+            HttpServletResponse response) {
+        getDatastream(pidString, null, download, request, response);
     }
 
     @RequestMapping("/file/{pid}/{datastream}")
-    public void getDatastream(@PathVariable("pid") String pid, @PathVariable("datastream") String datastream,
+    public void getDatastream(@PathVariable("pid") String pidString,
+            @PathVariable("datastream") String datastream,
             @RequestParam(value = "dl", defaultValue = "false") boolean download,
-            HttpServletRequest request, HttpServletResponse response) {
-        fedoraContentService.streamData(pid, datastream, download, new AnalyticsUserData(request), response);
+            HttpServletRequest request,
+            HttpServletResponse response) {
+
+        PID pid = PIDs.get(pidString);
+        AccessGroupSet principals = getAgentPrincipals().getPrincipals();
+
+        try {
+            fedoraContentService.streamData(pid, datastream, principals, download, response);
+            recordDownloadEvent(pid, datastream, principals, request);
+        } catch (IOException e) {
+            log.error("Problem retrieving {} for {}", pid.toString(), datastream, e);
+        }
+    }
+
+    private void recordDownloadEvent(PID pid, String datastream, AccessGroupSet principals,
+            HttpServletRequest request) {
+        if (!(StringUtils.isBlank(datastream) || ORIGINAL_FILE.equals(datastream))) {
+            return;
+        }
+        analyticsTracker.trackEvent(request, "download", pid, principals);
     }
 
     @RequestMapping("/thumb/{pid}")
     public void getThumbnailSmall(@PathVariable("pid") String pid,
             @RequestParam(value = "size", defaultValue = "small") String size, HttpServletRequest request,
             HttpServletResponse response) {
-        fedoraContentService.streamData(pid, ContentModelHelper.Datastream.THUMB_SMALL.getName(),
-                false, null, response);
+        // TODO implement retrieval of derivatives
```

Comment:
which JIRA ticket tracks this TODO?

Reply:
is it https://jira.lib.unc.edu/browse/BXC-2154?

Reply:
I can see jira at the moment, but the ticket is in the current sprint